sys/vfs_readdir.c

```diff
@@ -0,0 +1,41 @@
+#include <common.h>
+#include <dirent.h>
+#include <malloc.h>
+#include <uio.h>
+#include <vfs.h>
+
+int readdir_generic(vnode_t *v, uio_t *uio, readdir_ops_t *ops) {
+  void *it = ops->first(v);
+  dirent_t *dir;
+  off_t offset = 0;
+  int error;
+
+  /* Locate proper directory based on offset */
+  for (; it; it = ops->next(it)) {
+    unsigned reclen = _DIRENT_RECLEN(dir, ops->namlen_of(it));
+    if (offset + reclen > uio->uio_offset) {
+      assert(it == NULL || offset == uio->uio_offset);
+      break;
+    }
+    offset += reclen;
+  }
```

Comment:
```
     The getdirentries() system	call writes the	position of the	block read
     into the location pointed to by basep.  Alternatively, the	current	posi-
     tion pointer may be set and retrieved by lseek(2).	 The current position
     pointer should only be set	to a value returned by lseek(2), a value
     returned in the location pointed to by basep (getdirentries() only) or
     zero.
```

From my point of view this means we can use this argument the way we want. I was also confused because basep value on my linux machine has big integers looking to be quite random numbers, so I assumed they use it to point to data structures.

I've just read the `tmpfs_readdir` code in NetBSD (FreeBSD version is terrible to read) and it seems to me that they use uio_offset to store a sequence number, and based on that find directory entry. 
[tmpfs_readdir](http://bxr.su/NetBSD/sys/fs/tmpfs/tmpfs_subr.c#807)

They keep last read directory entry in cache:
[tn_readdir_lastp](http://bxr.su/NetBSD/sys/fs/tmpfs/tmpfs.h#145)

We could introduce caching to our interface somehow.

I'm surprised that NetBSD keeps cache inside tmpfs structure instead of file_t, I can't come up with reason they do that.